localparticipant.go

```diff
@@ -674,3 +678,176 @@ func (p *LocalParticipant) cleanup() {
 	p.rpcPendingAcks = &sync.Map{}
 	p.rpcPendingResponses = &sync.Map{}
 }
+
+func (p *LocalParticipant) StreamText(options StreamTextOptions) (*TextStreamWriter, error) {
+	if options.StreamId == nil {
+		streamId := uuid.New().String()
+		options.StreamId = &streamId
+	}
+
+	if options.Attributes == nil {
+		options.Attributes = make(map[string]string)
+	}
+
+	info := TextStreamInfo{
+		baseStreamInfo: &baseStreamInfo{
+			Id:         *options.StreamId,
+			MimeType:   "text/plain",
+			Topic:      options.Topic,
+			Timestamp:  time.Now().UnixMilli(),
+			Size:       options.TotalSize,
+			Attributes: options.Attributes,
+		},
+	}
+
+	header := &livekit.DataStream_Header{
+		StreamId:    info.Id,
+		MimeType:    info.MimeType,
+		Topic:       info.Topic,
+		Timestamp:   info.Timestamp,
+		TotalLength: info.Size,
+		ContentHeader: &livekit.DataStream_Header_TextHeader{
+			TextHeader: &livekit.DataStream_TextHeader{
+				OperationType: livekit.DataStream_CREATE,
+			},
+		},
+	}
+	if options.ReplyToStreamId != nil {
+		if textHeader, ok := header.ContentHeader.(*livekit.DataStream_Header_TextHeader); ok {
+			textHeader.TextHeader.ReplyToStreamId = *options.ReplyToStreamId
+		}
+	}
+
+	writer, err := newTextStreamWriter(info, header, p.engine, options.DestinationIdentities, options.OnProgress)
+	if err != nil {
+		return nil, err
+	}
+	p.engine.OnClose(func() {
+		err := writer.Close()
+		if err != nil {
+			logger.Errorw("error closing writer", err)
+		}
+	})
+
+	return writer, nil
+}
+
+func (p *LocalParticipant) SendText(text string, options StreamTextOptions) (*TextStreamInfo, error) {
+	if options.TotalSize == nil {
+		textInBytes := []byte(text)
+		totalTextLength := uint64(len(textInBytes))
+		options.TotalSize = &totalTextLength
+	}
+
+	writer, err := p.StreamText(options)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		// should return the error from here if any
+		_, err := writer.Write(text)
+		if err != nil {
+			logger.Errorw("error writing text", err)
+		}
+		err = writer.Close()
+		if err != nil {
+			logger.Errorw("error closing writer", err)
+		}
+	}()
+	return &writer.Info, nil
```

Comment:
How is the caller expected to use this return value? What if there is an error in writing (guessing closing error is not problematic)? Should the application retry? Is that what the info return value will be used for? In the example, I see that the return value is not used at all.

Have to use callback to notify errors when doing async send. Or use await semantics. There is a comment there about needing to return error, is that a TODO?

Reply:
yeah that is a todo, wasn't sure how to propagate the error without blocking anything. maybe callback is the way.